Sources/Testing/Testing.docc/testing-for-errors-in-swift-code.md

```diff
@@ -0,0 +1,131 @@
+throwing an error.  Use overrides of the `#expect()` and
+``require(_:_:sourceLocation:)-5l63q`` macros that check for errors.
+
+### Validate that your code throws an expected error
+
+The Swift structure in this example represents a list that accepts any
+number of toppings for pizzas in the list.  The API contains a method for
+applying a topping to a range of pizzas, and a method for retrieving the
+toppings requested for the item at a given index.  Both of these methods
+throw errors if their parameters are outside the list's range.
+
+```swift
+enum PizzaBase {
+    case deepCrust
+    case shallowCrust
+    case calzone
+}
+
+enum Topping {
+    case tomato
+    case cheese
+    case caper
+    case anchovy
+    case prosciutto
+    case pineapple
+}
+
+struct PizzaToppings {
+    enum PizzaToppingsError : Error {
+        case outOfRange
+    }
+
+    let pizzas: [PizzaBase]
+    var toppings: [Int: [Topping]]
+
+    init(bases: [PizzaBase]) {
+        pizzas = bases
+        toppings = [Int: [Topping]]()
+    }
+
+    mutating func add(topping: Topping, toPizzasIn range: Range<Int>) throws {
+        guard Int(range.startIndex) >= 0 && Int(range.endIndex) < pizzas.count else {
+            throw PizzaToppingsError.outOfRange
+        }
+        for index in range {
+            if var toppingList = toppings[index] {
+                toppingList.append(topping)
+                toppings[index] = toppingList
+            } else {
+                toppings[index] = [topping]
+            }
+        }
+    }
+
+    func toppings(forPizzaAt index: Int) throws -> [Topping] {
+        guard index >= 0 && index < pizzas.count else {
+            throw PizzaToppingsError.outOfRange
+        }
+        return toppings[index] ?? []
+    }
+
+    // Other methods.
+}
+```
+
+Create a test function that `throws` and `try` the code under test.
+If the code throws an error, then your test fails.
+
+To check that the code under test throws a specific error, or to continue a
+longer test function after the code throws an error, pass that error as the
+first argument of ``expect(throws:_:sourcelocation:performing:)-1xr34``, and
+pass a closure that calls the code under test:
+
+```swift
+@Test func cannotAddToppingToPizzaBeforeStartOfList() {
+    var order = PizzaToppings(bases: [.calzone, .deepCrust])
+    #expect(throws: PizzaToppings.PizzaToppingsError.outOfRange) {
+        try order.add(topping: .mozarella, toPizzasIn: -1..<0)
+    }
+}
+```
+
+If the closure completes without throwing an error, the testing library
+records an issue.  Other overloads of ``expect(_:_:sourceLocation:)`` let you test that
+the code throws an error of a given type, or matching an arbitrary
+Boolean test.  Similar overloads of ``require(_:_:sourceLocation:)-5l63q`` stop running your
+test if the code doesn't throw the expected error.
+
+### Validate that your code doesn't throw an error
+
+Validate that the code under test doesn't throw an error by comparing
```

Comment:
We should emphasize here that a test function that throws will fail if an error is thrown, so _usually_ you don't need to use `#expect(throws: Never.self)`. You'd really only use it if you want to record an issue for a thrown error but _not_ stop execution of the test function. It's pretty nuanced, to the point that I'm not sure we should even call it out in the documentation.

Reply:
There's some history here that we don't need to document, but for your own edification/amusement: XCTest in Objective-C has `XCTAssertNoThrow()` that asserts a given expression doesn't throw an _Objective-C or C++ exception_. This is different from a Swift error because if an exception is thrown through a test method's frame, it tends to wreak some havoc on the way out before being caught by XCTest, so you want to proactively catch them if you know they might occur. This was inherited as `XCTAssertNoThrow()` in Swift which checks if a function throws a Swift error rather than an exception.